src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -2153,6 +2154,190 @@ struct ConcatToBroadcast final
+  // purposes to use abs(frac(x)) = abs(x) - floor(abs(x)).
+  //
+  // Furthermore, pi * abs(frac(x)) loses precision when abs(frac(x)) is close
+  // to 1. To remedy this, we can use the fact that sin(pi * x) in the domain
+  // [0, 1] is symmetric across the line Y=0.5.
+  //
+
+  // Convert values of abs_frac > 0.5 to (1 - abs_frac) to improve precision of
+  // pi * abs_frac for values of abs_frac close to 1.
+  Value abs = rewriter.create<mlir::stablehlo::AbsOp>(loc, x);
+  Value absFrac = rewriter.create<mlir::stablehlo::SubtractOp>(
+      loc, abs, rewriter.create<mlir::stablehlo::FloorOp>(loc, abs));
+  Value reduceAbsFrac = rewriter.create<mlir::stablehlo::CompareOp>(
+      loc, half, absFrac, mlir::stablehlo::ComparisonDirection::LT);
+  absFrac = rewriter.create<mlir::stablehlo::SelectOp>(
+      loc, reduceAbsFrac,
+      rewriter.create<mlir::stablehlo::SubtractOp>(loc, one, absFrac), absFrac);
+
+  // Materialize reflection.
+  Value reflectionDenom = rewriter.create<mlir::stablehlo::LogOp>(
+      loc,
+      rewriter.create<mlir::stablehlo::SineOp>(
+          loc, rewriter.create<mlir::stablehlo::MulOp>(
+                   loc, mlir::stablehlo::getConstantLike(rewriter, loc, M_PI, x), absFrac)));
+  Value lgammaReflection = rewriter.create<mlir::stablehlo::SubtractOp>(
+      loc,
+      rewriter.create<mlir::stablehlo::SubtractOp>(
+          loc, mlir::stablehlo::getConstantLike(rewriter, loc, std::log(M_PI), x),
+          reflectionDenom),
+      lgamma);
+
+  // Avoid computing -inf - inf, which is nan. If reflection_denom is +/-inf,
+  // then it "wins" and the result is +/-inf.
+  Value finiteReflectionDenom =
+      rewriter.create<mlir::stablehlo::IsFiniteOp>(loc, reflectionDenom);
+  Value negReflectionDenom =
+      rewriter.create<mlir::stablehlo::NegOp>(loc, reflectionDenom);
+  lgammaReflection = rewriter.create<mlir::stablehlo::SelectOp>(
+      loc, finiteReflectionDenom, lgammaReflection, negReflectionDenom);
+
+  // Select whether or not to rely on the reflection.
+  lgamma = rewriter.create<mlir::stablehlo::SelectOp>(loc, needToReflect,
+                                                      lgammaReflection, lgamma);
+
+  // Materialize +/-inf behavior as
+  //   lgamma(+/-inf) = +inf.
+  Value xIsInf = rewriter.create<chlo::IsInfOp>(loc, x);
+  return rewriter.create<mlir::stablehlo::SelectOp>(
+      loc, xIsInf,
+      getConstantLikeInfValue(rewriter, loc, x, /*negative=*/false), lgamma);
+}
+
+struct GammaConstProp final : OpRewritePattern<mlir::chlo::LgammaOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::chlo::LgammaOp op,
+                                PatternRewriter &rewriter) const override {
+    // return if not constant
+    DenseElementsAttr inputAttr;
+    if(!matchPattern(op.getOperand(),m_Constant(&inputAttr)))
+      return failure();
+
+    Value result = materializeLgamma(rewriter,op.getLoc(),op->getOperands());
```

Comment:
so this is fun, but stablehlo at least has a eval method with the literal constant implemented. Apparently stablehlo doesn;t =/.

Since this code looks like it comes from lowerchlotostablehlo, maybe we can just run the relevant lowerchlo function here if it has a constant operand (rather than copying the lowering here).

Reply:
would that mean calling this?
https://github.com/openxla/stablehlo/blob/ef176a130f28196dcb4a5735d0f2f6ed0f85bd5d/stablehlo/transforms/ChloLegalizeToStablehlo.cpp#L1753

Reply:
yeah, but now that I look at it, it's marked static

Reply:
Would be happy to expose the functions in some shareable way! Could make a header `ChloDecompositionUtils.h` and can expose the individual decomp pattern, or the materialize function of interest.

Reply:
That would be helpful!